mailpoet/lib/Automation/Engine/Builder/DuplicateWorkflowController.php

```diff
@@ -0,0 +1,91 @@
+<?php declare(strict_types = 1);
+
+namespace MailPoet\Automation\Engine\Builder;
+
+use MailPoet\Automation\Engine\Data\NextStep;
+use MailPoet\Automation\Engine\Data\Step;
+use MailPoet\Automation\Engine\Data\Workflow;
+use MailPoet\Automation\Engine\Exceptions;
+use MailPoet\Automation\Engine\Exceptions\InvalidStateException;
+use MailPoet\Automation\Engine\Storage\WorkflowStorage;
+use MailPoet\Automation\Engine\WordPress;
+use MailPoet\Util\Security;
+
+class DuplicateWorkflowController {
+  /** @var WordPress */
+  private $wordPress;
+
+  /** @var WorkflowStorage */
+  private $workflowStorage;
+
+  public function __construct(
+    WordPress $wordPress,
+    WorkflowStorage $workflowStorage
+  ) {
+    $this->wordPress = $wordPress;
+    $this->workflowStorage = $workflowStorage;
+  }
+
+  public function duplicateWorkflow(int $id): Workflow {
+    $workflow = $this->workflowStorage->getWorkflow($id);
+    if (!$workflow) {
+      throw Exceptions::workflowNotFound($id);
+    }
+
+    $duplicate = new Workflow(
+      $this->getName($workflow->getName()),
+      $this->getSteps($workflow->getSteps()),
+      $this->wordPress->wpGetCurrentUser()
```

Comment:
I think that is a good move, but wanted to highlight, that we do not copy the author, but transfer the authorship (potentially later on maybe CRUD capabilities) to the current user.

Reply:
Yeah, [I thought this would be the correct behavior](https://github.com/mailpoet/mailpoet/pull/4376#discussion_r994554485). The person who duplicates a workflow is the one who creates the new one, I guess.